### PR TITLE
Add R package qgg

### PR DESCRIPTION
**Pull request by psoerensen**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by danielnachun** [comment]
Thanks for this submission.  Can you try to generate this recipe using `conda skeleton cran qgg`?  This will create a recipe that is compliant with what we need for Conda-Forge.  There's some comments we'll need to clean up afterward but the skeleton will cover most of what is needed.

**Comment by psoerensen** [comment]
Hi Daniel, I have updated the recipe using "conda skeleton cran qgg".
Please let me know if there are other things I should change. Kind regards
Peter

File: recipes/r-qgg/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    - r-mass
+    - r-mcmcpack
+    - r-matrix
+    - r-rcpp >=1.0.5
+    - r-rcpparmadillo
+    - r-coda
+    - r-corpcor
+    - r-data.table
+    - r-statmod
+    - {{ stdlib('c') }}
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:64]:
```suggestion
```
This is only needed in the `build` requirements

File: recipes/r-qgg/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    - libopenblas 
+    - libgfortran
+    
+  run:
+    - r-base
+    - {{native}}gcc-libs         # [win]
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:70]:
```suggestion
```
This no longer needed - `conda skeleton cran` is a little outdated now but deleting this should allow your Windows build to progress, hopefully fixing it.

File: recipes/r-qgg/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    - r-mass
+    - r-mcmcpack
+    - r-matrix
+    - r-rcpp >=1.0.5
+    - r-rcpparmadillo
+    - r-coda
+    - r-corpcor
+    - r-data.table
+    - r-statmod
+    - {{ stdlib('c') }}
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:80]:
```suggestion
```
This is only needed in `build` requirements.

File: recipes/r-qgg/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    - libopenblas
+    - libgfortran
+
+test:
+  commands:
+    # You can put additional test commands to be run here.
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:86]:
```suggestion
```
Most of these comments are just helpers that can be deleted now - I'll make a suggestion for that ones that should be deleted.

File: recipes/r-qgg/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    - $R -e "library('qgg')"           # [not win]
+    - "\"%R%\" -e \"library('qgg')\""  # [win]
+
+  # You can also put a file called run_test.py, run_test.sh, or run_test.bat
+  # in the recipe that will be run at test time.
+
+  # requires:
+    # Put any additional test requirements here.
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:94]:
```suggestion
```

File: recipes/r-qgg/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    - {{ compiler('cxx') }}            # [not win]
+    - {{ compiler('m2w64_cxx') }}      # [win]
+    - {{ compiler('fortran') }}        # [not win]
+    - {{ compiler('m2w64_fortran') }}  # [win]
+    - {{ posix }}filesystem        # [win]
+    - {{ posix }}make
+    - {{ posix }}sed               # [win]
+    - {{ posix }}coreutils         # [win]
+    - {{ posix }}zip               # [win]
+    - {{ stdlib("c") }}
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:51]:
```suggestion
    - {{ stdlib("c") }}
    - {{ stdlib("m2w64_c") }}
```
We need `m2w64_c` for Windows here.

File: recipes/r-qgg/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+source:
+  url:
+    - {{ cran_mirror }}/src/contrib/qgg_{{ version }}.tar.gz
+    - {{ cran_mirror }}/src/contrib/Archive/qgg/qgg_{{ version }}.tar.gz
+  sha256: dff06a1f4684622a4e60f617758254d25210f5081d814bc143906cea8b297a19
+
+build:
+  merge_build_host: True  # [win]
+  # If this is a new build for the same version, increment the build number.
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:18]:
```suggestion
```
`merge_build_host` should no longer be needed on Windows

File: recipes/r-qgg/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,138 @@
+{% set version = '1.1.6' %}
+
+{% set posix = 'm2-' if win else '' %}
+{% set native = 'm2w64-' if win else '' %}
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:4]:
```suggestion
```
This is no longer needed

File: recipes/r-qgg/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+about:
+  home: https://github.com/psoerensen/qgg
+  license: GPL-3.0-or-later
+  summary: 'Provides an infrastructure for efficient processing of large-scale genetic and phenotypic
+    data including core functions for: 1) fitting linear mixed models, 2) constructing
+    marker-based genomic relationship matrices, 3) estimating genetic parameters (heritability
+    and correlation), 4) performing genomic prediction and genetic risk profiling, and
+    5) single or multi-marker association analyses. Rohde et al. (2019) <doi:10.1101/503631>.'
+  license_family: GPL3
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:104]:
```suggestion
```
This is no longer needed.

File: recipes/r-qgg/build.sh [hunk 10]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+
+# 'Autobrew' is being used by more and more packages these days
+# to grab static libraries from Homebrew bottles. These bottles
+# are fetched via Homebrew's --force-bottle option which grabs
+# a bottle for the build machine which may not be macOS 10.9.
+# Also, we want to use conda packages (and shared libraries) for
+# these 'system' dependencies. See:
+# https://github.com/jeroen/autobrew/issues/3
```
Comment (danielnachun) [inline recipes/r-qgg/build.sh:9]:
```suggestion
```

File: recipes/r-qgg/build.sh [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+mv DESCRIPTION DESCRIPTION.old
+grep -va '^Priority: ' DESCRIPTION.old > DESCRIPTION
+# shellcheck disable=SC2086
+${R} CMD INSTALL --build . ${R_ARGS}
+
+# Add more build steps here, if they are necessary.
+
+# See
+# https://docs.conda.io/projects/conda-build
+# for a list of environment variables that are set during the build process.
```
Comment (danielnachun) [inline recipes/r-qgg/build.sh:22]:
```suggestion
```

File: recipes/r-qgg/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+  url:
+    - {{ cran_mirror }}/src/contrib/qgg_{{ version }}.tar.gz
+    - {{ cran_mirror }}/src/contrib/Archive/qgg/qgg_{{ version }}.tar.gz
+  sha256: dff06a1f4684622a4e60f617758254d25210f5081d814bc143906cea8b297a19
+
+build:
+  merge_build_host: True  # [win]
+  # If this is a new build for the same version, increment the build number.
+  number: 0
+  # no skip
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:20]:
```suggestion
```

File: recipes/r-qgg/meta.yaml [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+  # This is required to make R link correctly on Linux.
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:22]:
```suggestion
```

**Review by danielnachun — CHANGES_REQUESTED** [review]
I've made some suggestions. Mostly to clean up some cruft from the recipe generator, as well as a fix for Windows.  Let's see after this the Windows build passes and then we should be good to go.

**Comment by psoerensen** [comment]
Not sure if I have done the right thing - still issues with the builds

File: recipes/r-qgg/meta.yaml [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    - r-mcmcpack
+    - r-matrix
+    - r-rcpp >=1.0.5
+    - r-rcpparmadillo
+    - r-coda
+    - r-corpcor
+    - r-data.table
+    - r-statmod
+    - libopenblas 
+    - libgfortran
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:61]:
```suggestion
    - libgfortran
```
I think this is no longer needed either, and is one of the reasons for some of the failures

File: recipes/r-qgg/meta.yaml [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+    - {{ compiler('m2w64_cxx') }}      # [win]
+    - {{ compiler('fortran') }}        # [not win]
+    - {{ compiler('m2w64_fortran') }}  # [win]
+    - {{ posix }}filesystem        # [win]
+    - {{ posix }}make
+    - {{ posix }}sed               # [win]
+    - {{ posix }}coreutils         # [win]
+    - {{ posix }}zip               # [win]
+    - {{ stdlib("c") }}
+    - {{ stdlib("m2w64_c") }}
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:47]:
```suggestion
    - {{ stdlib("c") }}  # [not win]
    - {{ stdlib("m2w64_c") }}  #[win]
```
I forgot the selectors here, sorry about that.

File: recipes/r-qgg/meta.yaml [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+    - r-mcmcpack
+    - r-matrix
+    - r-rcpp >=1.0.5
+    - r-rcpparmadillo
+    - r-coda
+    - r-corpcor
+    - r-data.table
+    - r-statmod
+    - libopenblas
+    - libgfortran
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:75]:
```suggestion
```
See above that this shouldn't be needed.

File: recipes/r-qgg/meta.yaml [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+    - r-mass
+    - r-mcmcpack
+    - r-matrix
+    - r-rcpp >=1.0.5
+    - r-rcpparmadillo
+    - r-coda
+    - r-corpcor
+    - r-data.table
+    - r-statmod
+    - libopenblas
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:60]: This may also be unneeded, but let's see what happens after you fix the other issues first.

**Comment by psoerensen** [comment]
I made additional changes with a bit of "trial and error" and search in other recipes. Check passes, but I am not sure if things are OK.

File: recipes/r-qgg/meta.yaml [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+    - {{ compiler('fortran') }}        # [not win]
+    - {{ compiler('m2w64_fortran') }}  # [win]
+    - {{ posix }}filesystem        # [win]
+    - {{ posix }}make
+    - {{ posix }}sed               # [win]
+    - {{ posix }}coreutils         # [win]
+    - {{ posix }}zip               # [win]
+
+  host:
+    - libopenblas
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:54]:
```suggestion
```
This should also be brought in through `r-base`, so we can remove it here.

File: recipes/r-qgg/meta.yaml [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+    - r-rcpp >=1.0.5
+    - r-rcpparmadillo
+    - r-coda
+    - r-corpcor
+    - r-data.table
+    - r-statmod
+
+  run:
+    - libopenblas
+    - m2w64-gcc-libgfortran  # [win]
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:68]:
```suggestion
```
This should be added automatically by `{{ compiler('m2w4_fortran') }}` - let's remove it and see what happens.

File: recipes/r-qgg/meta.yaml [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+    - r-matrix
+    - r-rcpp >=1.0.5
+    - r-rcpparmadillo
+    - r-coda
+    - r-corpcor
+    - r-data.table
+    - r-statmod
+
+  run:
+    - libopenblas
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:67]:
```suggestion
```
This should be unneeded - see above comment.

File: recipes/r-qgg/meta.yaml [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+#    - r-corpcor
+#    - r-rcpparmadillo
+#    - r-matrix
+#    - r-rcpp
+#    - r-statmod
+#    - r-mass
+#    - r-mcmcpack
+#    - r-coda
+
+
```
Comment (danielnachun) [inline recipes/r-qgg/meta.yaml:35]:
```suggestion
```
Although `conda-build` reports these as unneeded, they are definitely necessary.  The output is unfortunately a bit confusing.

**Review by danielnachun — COMMENTED** [review]
The recipe is in great shape - I've just suggested a few small cleanups.  Generally for R recipes we try to let `r-base` and the `compiler` packages handle all the dependencies which are not R packages.  As long as the build passes after this cleanup, we should be all set!

**Comment by psoerensen** [comment]
I made the suggested changes. However check fails now. I had added these arguments in order to get them to pass. Please let me know if you have alternative suggestions to get it to pass.

**Comment by danielnachun** [comment]
Let's put back what you had - there are some issues this may be uncovering relating to run exports, but these need to be fixed in other packages, not here.  Once that's done we'll be all set here.

**Comment by psoerensen** [comment]
Added back and checks pass again. Looking forward to understand what is happening. Thanks for the help!

**Comment by danielnachun** [comment]
Great work @psoerensen!  @conda-forge/staged-recipes @conda-forge/help-r This is ready for review.

**Comment by psoerensen** [comment]
@conda-forge/help-r @conda-forge/help-c-cpp @conda-forge/help-cdts

This recipe contains an R package (`qgg`) that includes both C++ and Fortran code. Please review. Thank you!

**Review by mfansler — CHANGES_REQUESTED** [review]
Thanks for submitting this! ❤️ 

There are a few things to tidy up before merging.

@danielnachun thanks for helping with this. A couple notes:
  - `conda_r_skeleton_helper` is the canonical way to generate recipes for CRAN packages;
  - until R migration is closed out, `staged-recipes` Windows builds are pre-UCRT and therefore might still need the `native` variable, build-host merge, etc.. Often easiest is just to recommend skipping on Windows and postpone troubleshooting until the UCRT migration hits the feedstock.

File: recipes/r-qgg/meta.yaml [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+    - r-statmod
+
+test:
+  commands:
+    - $R -e "library('qgg')"           # [not win]
+    - "\"%R%\" -e \"library('qgg')\""  # [win]
+
+about:
+  home: https://github.com/psoerensen/qgg
+  license: GPL-3.0-or-later
```
[the remark this answers is not in the document]
Comment (mfansler) [reply recipes/r-qgg/meta.yaml:75]: @psoerensen just noticed your the author 🤦  So if you want the v3+ here, feel free to ignore this suggestion.

**Review by mfansler — APPROVED** [review]
Looks good!

**Pull request merged**